### DISCUSSION TITLE
Apple ][ version

### DISCUSSION THREAD

**Discussion by SEGStriker** [Dietrich-L/CPM-65#1] [Ideas]
Hello,

Can you make this compatible with Apple ][?
Also, do you have an emulator that can be used to run your OS?

Thanks,
Stan

**Comment by Dietrich-L** [comment]
Excellent idea. I never thought abaout that. It should work though. I happen to own a working Apple II with a Floppy-EMU from BMOW. However I am by no means an Apple expert. We would need to write a suitable BIOS based on Apple firmware.
Let me think about it for a couple of days to devise a suitable way how to proceed.
Would you be interested in a collaboration effort?
Also I have no experience with 6502 emulators. Could you point me in a suitable direction?

Lets have a look at this

Dietrich

**Reply by SEGStriker** [reply]
What do you need? CPU only emulator or full one like Applewin? There are more, depending of which OS are you using mostly.
But you have working Apple II, so an emulator is not needed, except if you just want to make the developing work easier.
You can check this site about CPU emulators: http://www.6502.org/tools/emu
I have a lot of experience with Applewin and various other emulators, but I can't recommend which 6502 emulator will be most suitable to you. Just read their description, if you really need such.
I can recommend to you one really interesting OS for Apple II, which was written by 12 years old Bulgarian (in those days), later he wrote his onw emulator - Apple Oasis. It is basically something like MS-DOS (not CP/M) for Apple II: https://ftp.apple.asimov.net/images/masters/other_os/USDOS.dsk
On this site you can find a lot of software for Apple II, also ROMs, emulators, manuals, books, documentation and whatever you can imagine.
Let me know if I can help with something. I am not a programmer, even though I can program on several languages and my career is in IT - System Engineer, Cloud Engineer, System Administrator, etc.

**Reply by Dietrich-L** [reply]
We would need a 6502 simulator for preferably WIN10 which has also chr i/o and block device i/o simulation. So pure 6502 simulators are not very useful, because they lack these. Based on those features, the development of a first BIOS should be easy.

But on the other hand, doing it on an Apple II might be more straight forward. I will have to look in the handbooks for memory map, chr i/o and disk i/o.

Dietrich

**Comment by SEGStriker** [comment]
Here is some useful documentation about A2:

https://ftp.apple.asimov.net/documentation/os/dos - DOS
https://ftp.apple.asimov.net/documentation/os/cpm - Apple CP/M (with Z80 softcard adapter)
https://ftp.apple.asimov.net/documentation/os/prodos - ProDOS
https://ftp.apple.asimov.net/documentation/hardware - Hardware (including schematics)
https://ftp.apple.asimov.net/documentation/misc - Technical & programming references, etc
https://ftp.apple.asimov.net/documentation/programming - Programming languages for A2
https://gswv.apple2.org.za/a2zine/GS.WorldView/Resources/GS.TECH.INFO/MEGA.PEEKS.AND.POKES.html - List of important addresses
http://www.myoldmac.org/FAQ/Apple-II_Peek_Poke_Call.html - Another list

I knew a lot of such info, now a little rusty, but I still know where to find them.
We can use something like Skype to speak about how can I help further.

**Reply by Dietrich-L** [reply]
Thanks for the links.
I believe that [https://ftp.apple.asimov.net/documentation/os/cpm/Arnow_The_Apple_CPM_Book_1985.pdf](url) might also be useful.

Basically we need 2 things
1. A primitive BIOS with the most important routines for chr i/o and disk sector R/W - probably easy and straight forward
2. A boot routine to place the whole system into RAM at the correct locations and start it - hmm, there I have no expertise

I will prepare some docs to make things more transparent and come back to you.

Dietrich

**Reply by SEGStriker** [reply]
Yeah, that CP/M and others there could be useful, but, the Apple CP/M is somewhat unique. It uses Z80 CPU that is on an adapter (on some computers it is onboard, but still mapped to slot 4 - examples Pravetz-8M, Basis 108). However, it still uses 6502 CPU for the I/O operations, disk activity, etc. So it differs a lot from ordinary CP/M, but still it is very different from completely 6502 implementation. That's why I sent you that link to US-DOS, which is basically 6502 OS, looks very much like MS-DOS (and CP/M of course), it already implemented many of these challenges.

**Reply by Dietrich-L** [reply]
I tried to open UDOS.DSK with Ciderpress, but unfortunately it doesn't recognize the format. Any hints are appreciated.
Do you have sources for this system?

Dietrich

**Reply by SEGStriker** [reply]
I see, it cannot open it (another disk format), but there are tools to exchange files between DOS 3.3 and US-DOS. You can open it in Applewin and try, or you can transfer it to real disk and start it in your Apple II. Also, because you said you have floppy-emu, you can just copy the .dsk there and start it from the SD card.
About the source code, I don't have it, but it can be disassembled easily. And we have this, which could also help a lot: https://computerhistory.org/blog/apple-ii-dos-source-code

**Reply by Dietrich-L** [reply]
Success. My Apple reads and starts the image. I could format a disk and copy a file onto it. So far so good.

Regarding disassembly - that would be a way to realize the migration of CPM-65 to the Apple II very easily, because MSDOS and  CPM are very close relatives. However I have no experience in disassembling code on the Apple. Could you do that and create a disassembly txt file?
I then would open a separate space on my GITHUB, where we could work together on this project more easily.

Dietrich

**Reply by Dietrich-L** [reply]
I played a bit with USDOS. The basics work. However I could not make TRANSFER to work - always FILE NOT FOUND error. The help file is bulgarian only. So no luck there. I don't see a way to transfer the disk in a more readable format on my PC. Do you have any ideas?

**Comment by SEGStriker** [comment]
It probably transfer a file from DOS 3.3 to US DOS. And help command doesn't display anything about transfer. It works actually, also CATALOG - you have to change the disk, so far I don't see a way to force it to use drive 2.
I contacted the author and I am waiting for his response. Maybe he would like to contribute his source, if he still has it. Or he could join us, as an active programmer, who wrote OS, Apple Oasis emulator and many related tools, he would be very valuable, if he agree.

**Reply by Dietrich-L** [reply]
#It probably transfer a file from DOS 3.3 to US DOS. #. 
I tried that - didn‘t work either. Having the USDOS author on board would be wonderful.

**Reply by SEGStriker** [reply]
Let see if he will agree.
Meanwhile, here is an example how you can do a quick and dirty disassembly of anything that is in Apple II memory - using Applewin,
![Applewin_debug](https://user-images.githubusercontent.com/31370192/229631953-88fb9173-d736-411a-885f-6a90385e1855.JPG)

**Reply by Dietrich-L** [reply]
Ok. I have Applewin up and running. USDOS works, but can access drive A: only. I don't know why.
I also isolated the USDOS boot sector and disassembled it. Looks ok, but I know too little about the Apple DOS to push forward. Basically the boot sector should be enough to load the whole system to memory and start it. That is, what CPM-65 does on my Junior Computer.
So now I need a description of the Apple DOS RWTS - entry points, data to be supplied. Ideas?
And then I need the adresses of the console input and output routines in the Apple ROM and instructions how to switch the display to 80 coloumn mode.
And finally I would have to prepare an image disk with the new boot sector and CPM-65 BIOS, BDOS and CCP and the game is up.
Should be doable, but needs some work and foremost time ;-)

Dietrich

**Reply by SEGStriker** [reply]
https://github.com/cmosher01/Apple-II-Source/blob/master/src/dos/system/common/rwts.m4

Switching to 80 column mode is easy, but you have to check if there is 80 column card in slot 3 (AUX slot of //e is mapped there as well).

PR#3 from Basic/DOS

Machine language:

20 00 C3 - JSR $C300 (address of slot 3)
60 - RTS

After 80 column mode is activated, there are some ways to switch to 40 columns without turning off the 80 column card - Esc-4 from Basic/DOS (Esc-4 returns to 80 column mode). PR#0 (JSR $C000 or CALL $C000 from Basic) turns it off by restoring the output to slot 0 (default).

**Comment by Dietrich-L** [comment]
The Apple ][ port is available under https://github.com/Dietrich-L/CPM-65-for-Apple-II

**Comment by Dietrich-L** [comment]
Solved
